back/engines/commercial/machine_translations/app/services/machine_translations/machine_translation_service.rb

```diff
@@ -1,47 +1,66 @@
 module MachineTranslations
   class MachineTranslationService
-
     def build_translation_for translatable:, attribute_name:, locale_to:
       multiloc = translatable[attribute_name]
       locale_from = multiloc_to_locale_from multiloc
       text_or_html = multiloc[locale_from]
       MachineTranslation.new(
         translatable: translatable,
-        attribute_name: attribute_name, 
-        locale_to: locale_to, 
+        attribute_name: attribute_name,
+        locale_to: locale_to,
         translation: translate(text_or_html, locale_from, locale_to)
-        )
+      )
     end
 
-    def assign_new_translation translation
+    def assign_new_translation(translation)
       multiloc = translation.translatable[translation.attribute_name]
       locale_from = multiloc_to_locale_from multiloc
       text_or_html = multiloc[locale_from]
       translation.assign_attributes translation: translate(text_or_html, locale_from, translation.locale_to)
     end
 
-    def multiloc_to_locale_from multiloc
+    def multiloc_to_locale_from(multiloc)
       # It would be nice if we could figure out
       # the better locale to translate from.
       multiloc.keys.first
     end
 
-    def translate text_or_html, locale_from, locale_to, retries: 10
+    def translate(text_or_html, locale_from, locale_to, retries: 10)
       # Uses Google translate
       from = locale_from[0...2]
       to = locale_to[0...2]
-      if from == to
-        text_or_html
-      else
-        while retries > 0
-          begin
-            return EasyTranslate.translate(text_or_html, :from => from, :to => to)
-          rescue EasyTranslate::EasyTranslateException => e
-            retries -= 1
-            sleep rand(60)
-          end
+
+      return text_or_html if from == to
+
+      translation = nil
+      exception = nil
+      while retries > 0
+        begin
+          translation = EasyTranslate.translate text_or_html, from: from, to: to
+          break
+        rescue EasyTranslate::EasyTranslateException => e
+          exception = e
+          retries -= 1
+          sleep rand(60)
         end
       end
+
+      if translation.present?
+        fix_html translation
+      elsif exception.present?
+        raise exception
+      else
+        ''
+      end
+    end
+
+    private
+
+    def fix_html(text_or_html)
+      doc = Nokogiri::HTML.fragment text_or_html
+      return text_or_html if doc.text == text_or_html # plain text
+
+      doc.errors.any? ? doc.to_s : text_or_html
```

Comment:
This means that if there are errors, they are swept under the mat and whatever Nokogiri returns as `text` is fine. Is that the expected behaviour?

Reply:
Yes, it seems that google translate returns & instead of &amp; in translated HTML, and so the proposed solution is to let Nokogiri fix the HTML if it found something wrong with it. Do you think it would be useful to log the error in Sentry in those cases? I wonder if we wouldn't get exceptions with a high count if we would do this.